### PR TITLE
Port PathSpec implementation to JS

### PR DESCRIPTION
**Pull request by kalman**

### CODE REVIEW

**Comment by kalman** [comment]
Not ready for review, just testing out the PR builder.

**Comment by kalman** [comment]
PTAL.

**Comment by kalman** [comment]
This is the last one :-)

**Comment by kalman** [comment]
The comment in the patch is:

```
    This is a breaking change to the JS API. HashSpec and parseObjectSpec
    have been removed in favour of PathSpec, and parse errors now throw
    SyntaxError instead of silently failing (this matches JSON.parse).
```

File: js/src/specs.js [hunk 1]
```diff
@@ -4,155 +4,180 @@
...
-  static parse(spec: string): ?DatabaseSpec {
-    const match = spec.match(/^(.+?)(:.+)?$/);
-    if (!match) {
-      return null;
-    }
-    const [, scheme, p] = match;
-    if (p && p.indexOf('::') > 0) {
-      return null;
+  /** Cache this for testing, or it will reset every time database() is called. */
+  _ms: MemoryStore|null;
```
Comment (arv) [inline js/src/specs.js]: Whitespace around `|`

File: js/src/specs.js [hunk 2]
```diff
@@ -4,155 +4,180 @@
...
-        return null;
+        throw new SyntaxError(`Unsupported protocol ${protocol}`);
     }
-    return new this(scheme, path || '');
   }
 
   constructor(scheme: string, path: string) {
     this.scheme = scheme;
     this.path = path;
+    this._ms = scheme === 'mem' ? new MemoryStore() : null;
```
Comment (arv) [inline js/src/specs.js]: I think the comment would be better here.

File: js/src/specs.js [hunk 3]
```diff
@@ -4,155 +4,180 @@
...
+  /**
+   * Returns a tuple of the database backed by this dataset, and value at the HEAD of this
+   * dataset. If this dataset doesn't have a HEAD, the value will be `null`.
+   *
+   * The caller should always call `close()` when done.
+   */
+  value(): Promise<[Database, Value|null]> {
+    const db = this.database.database();
+    return this.dataset().head().then(commit => {
+      if (commit === null || commit === undefined) {
```
Comment (arv) [inline js/src/specs.js]: if (!commit) {

**Comment by arv** [comment]
Nice. You need to update some of the samples/js though.

File: js/src/specs.js [hunk 4]
```diff
@@ -4,155 +4,180 @@
...
-    // can't call close() on it. Ideally, we'd fix that.
-    return this.dataset().head()
-      .then(commit => commit && commit.value);
+  /**
+   * Returns a tuple of the database backed by this dataset, and value at the HEAD of this
+   * dataset. If this dataset doesn't have a HEAD, the value will be `null`.
+   *
+   * The caller should always call `close()` when done.
+   */
+  value(): Promise<[Database, Value|null]> {
```
Comment (arv) [inline js/src/specs.js]: This naming is not very idiomatic js but this is old code...

File: js/src/specs.js (hunk 3 above)
Comment (kalman) [inline js/src/specs.js]: done

File: js/src/specs.js [hunk 5]
```diff
@@ -4,155 +4,180 @@
...
-  // Returns parsed spec, or null if the spec was invalid.
-  static parse(spec: string): ?DatabaseSpec {
-    const match = spec.match(/^(.+?)(:.+)?$/);
-    if (!match) {
-      return null;
-    }
-    const [, scheme, p] = match;
-    if (p && p.indexOf('::') > 0) {
-      return null;
+  /** Cache this for testing, or it will reset every time database() is called. */
```
Comment (kalman) [inline js/src/specs.js]: done

File: js/src/specs.js (hunk 1 above)
Comment (kalman) [inline js/src/specs.js]: I like it like this :-) but fair enough. done.

File: js/src/specs.js (hunk 2 above)
Comment (kalman) [inline js/src/specs.js]: done

File: js/src/specs.js (hunk 4 above)
Comment (kalman) [inline js/src/specs.js]: what should it be?

Comment (arv) [inline js/src/specs.js]: Probably `getValue` since using getters for Promises is too weird.

Comment (kalman) [inline js/src/specs.js]: It's not a getter, but it does sound like a getter. I actually might as well change it.

Comment (kalman) [inline js/src/specs.js]: eh...

Comment (kalman) [inline js/src/specs.js]: It's a breaking change anyway (return type changed) so I *can* change it without any more breakage. Let me know if you want that.

**Comment by kalman** [comment]
PTAL.

File: js/src/specs.js (hunk 4 above)
Comment (arv) [inline js/src/specs.js]: Let's leave it as is for now. There are a few other of these already.

File: js/src/specs.js [hunk 6]
```diff
@@ -4,155 +4,175 @@
...
-  value(): Promise<any> {
-    const database = this.database.database();
-    return database.readValue(this.hash).then(v => database.close().then(() => v));
+  /**
+   * Resolves this PathSpec, and returns the database it was resolved in, and the value it
+   * resolved to. If the value wasn't found, it will be `null`.
+   *
+   * The caller should always call `close()` when done.
+   */
+  value(): Promise<[Database, Value|null]> {
```
Comment (arv) [inline js/src/specs.js]: I still think we should use spaces around `|` since all other binops require it.

**Comment by arv** [comment]
LGTM with nit

File: js/src/specs.js (hunk 6 above)
Comment (kalman) [inline js/src/specs.js]: :-(

**Pull request merged**